### ISSUE TITLE
Allow specifying a non-zero start angle for SpotLight3D (resulting in cylindrical lights)

### ISSUE THREAD

**Issue by Calinou** [godotengine/godot-proposals#2520]
*Related to https://github.com/godotengine/godot-proposals/issues/1275.*

### Describe the project you are working on

The Godot editor :slightly_smiling_face: 

### Describe the problem or limitation you are having in your project

SpotLights currently must have a conic shape, which makes them too limited to imitate DirectionalLight3Ds that are constrained to a specific area of a level.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Cylindrical lights (spot lights whose "start angle" is equal to the "end angle") can be used as localized DirectionalLight3D. Since SpotLight3D shadows are much cheaper to render than DirectionalLight3D shadows, this can improve performance in indoor levels and closed areas while avoiding potential issues with GI light leaks.

Being able to specify a start angle also makes it possible to fake area lights in a more convincing manner.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Add a `start_angle` property (defaulting to `0`) that specifies how wide the "circle" at the start of the SpotLight should be. If `start_angle` is equal to the end angle, then the spot light effectively becomes *cylindrical* and acts like a localized DirectionalLight3D.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No. While OmniLight3Ds and even SpotLight3Ds can be used to achieve this kind of "localized" DirectionalLight3D effect, light and shadow rendering will look incorrect unless the light is placed *really* far away. This greatly lowers the effective shadow resolution. OmniLight3Ds also have to render cubemap (or dual parabolid) shadow maps, which are more expensive than SpotLight3D shadows since they only have to render one face.

### Is there a reason why this should be core and not an add-on in the asset library?

See above.

*Keywords for easier searching: cylinder*

**Comment by alexfreyre** [comment]
Can get this type of light a square shape and a square specular lobe?

**Comment by Calinou** [comment]
> Can get this type of light a square shape and a square specular lobe?

That's probably quite different to implement, so I doubt it.

**Comment by golddotasksquestions** [comment]
Yes please! This proposal is exactly what I would have needed when I was making the blob shadow for the main character in [this demo](https://www.reddit.com/r/godot/comments/mcqzwc/almost_perfect_pixel_perfect_25d_pixelart_demo/).

**Comment by Calinou** [comment]
> Yes please! This proposal is exactly what I would have needed when I was making the blob shadow for the main character in [this demo](https://www.reddit.com/r/godot/comments/mcqzwc/almost_perfect_pixel_perfect_25d_pixelart_demo/). I can think of a million other cases where it would be useful.

For blob shadows, you should use Decal nodes (new in `master`) instead of negative lights.

**Comment by golddotasksquestions** [comment]
> For blob shadows, you should use Decal nodes (new in master) instead of negative lights. This gives you more control over the shadow's size and normal fade, while putting less strain on the renderer. Decal nodes can be moved every frame just fine, since they don't rely on mesh generation to be rendered.

Yes, decals would have been my first choice and I'm really looking forward to the 4.0 decals. The decal solutions available for Godot 3 did not work for me in this demo, regardless of what I tried. This was really just one example where an alternative like this would be awesome. But I can think of many more situations where non-zero  start angle for lights would be great. Moving the lights way out of bounds and then having to struggle with tons of cull masks and light layers which all have to be turned on and off in certain conditions is really not fun.

**Comment by jitspoe** [comment]
Oops, didn't notice this when I proposed https://github.com/godotengine/godot-proposals/issues/9157

I think this might be slightly different, though, or maybe it's just a terminology thing?  Like regardless of the angle, the start is still going to be a single point.  Wouldn't a cylindrical light be an angle of 0 with a starting radius?

On the topic of spot lights used like directional lights, I have a case where I'm using a very large spot light that's almost a directional light, and it would be nice if I could specify a separate texture for it instead of using the atlas so it's not so chunky.  Any proposals to have special non-atlased lights?

**Comment by Calinou** [comment]
> Wouldn't a cylindrical light be an angle of 0 with a starting radius?

That's a spotlight, which is how it works currently :slightly_smiling_face: 

A cylindrical light is a light whose start and end angle are identical.

> and it would be nice if I could specify a separate texture for it instead of using the atlas so it's not so chunky. Any proposals to have special non-atlased lights?

The way the shadow atlas works doesn't make this possible. However, there could be a way to force specific lights to use a given shadow quadrant, which can be used to increase the shadow detail of certain lights: https://github.com/godotengine/godot-proposals/issues/840

**Comment by jitspoe** [comment]
I think there's some disconnect here... spotlights have the same angle all the way through:

![image](https://github.com/godotengine/godot-proposals/assets/7751182/f1110dd7-fb20-45e8-beb7-d5c94836d949)

A cylinder would have an angle of 0 degrees + a radius.
![image](https://github.com/godotengine/godot-proposals/assets/7751182/0615c0a9-46ce-40a6-8bd8-286a279709f2)

**Comment by Calinou** [comment]
> I think there's some disconnect here... spotlights have the same angle all the way through:

By angle, I meant the radius of the circle rather than the angle.

**Comment by ben-rolfe** [comment]
I just want to add that this would be especially useful for moving lights (e.g. flashlights and car headlights). I'm currently working on a car game and it looks really silly when you drive up to a wall and the headlights become pinpoints - My only available workaround seems to be to embed the lights in the body of the car, and that causes its own problems.

**Comment by jitspoe** [comment]
Would this be as simple as adding a near clip for the shadow rendering and offsetting the light point back?  Hmm, no, that would potentially allow the spotlight to light stuff behind the source.

**Comment by andreasng** [comment]
a near clip would be more than enough. Would be a very nice feature and genrally comes standard in any other engine so i personally think it would be welcome,

**Comment by JD-The-65th** [comment]
+1 for this, if I knew C++ better, I would try to tackle this myself, but having this as a feature would massively benefit light realism when working with Volumetric Fog and Physically Based light systems. As far as I can tell, most lights don't just emit from a single 0 radius point. 

<img width="267" height="466" alt="Image" src="https://github.com/user-attachments/assets/40eff241-0043-4355-a064-e4904160cb2e" />
